models/demos/whisper/tt/ttnn_optimized_functional_whisper.py

```diff
@@ -0,0 +1,672 @@
+        )
+    else:
+        query_states, key_states, value_states = calculate_query_key_values(
+            config, hidden_states, parameters=parameters.query_key_value, whisper_memory_config=whisper_memory_config
+        )
+
+    query_states *= scaling
+    attn_weights = ttnn.matmul(query_states, key_states)
+
+    if attention_mask is not None:
+        attn_weights = ttnn.add(attn_weights, attention_mask)
+
+    # differences in ttnn.softmax vs torch.softmax cause the attn_weights to be slightly different
+    attn_weights = ttnn.softmax(attn_weights, dim=-1)
+
+    attn_probs = dropout(attn_weights, p=0, training=False)
+    attn_output = ttnn.matmul(attn_probs, value_states, memory_config=whisper_memory_config)
+
+    ttnn.deallocate(attn_probs)
+    ttnn.deallocate(attn_weights)
+    ttnn.deallocate(query_states)
+
+    attn_output = ttnn.transformer.concatenate_heads(attn_output)
+
+    attn_output = ttnn.linear(
+        attn_output,
+        parameters.out_proj.weight,
+        bias=parameters.out_proj.bias,
+        memory_config=whisper_memory_config,
+    )
+
+    return attn_output
+
+
+def encoder_layer(config, device, hidden_states, *, parameters, whisper_memory_config):
+    residual = hidden_states
+
+    hidden_states = ttnn.layer_norm(
+        hidden_states,
+        weight=parameters.self_attn_layer_norm.weight,
+        bias=parameters.self_attn_layer_norm.bias,
+        memory_config=whisper_memory_config,
+    )
+
+    hidden_states = whisper_attention(
+        config,
+        device,
+        hidden_states,
+        attention_mask=None,
+        parameters=parameters.self_attn,
+        whisper_memory_config=whisper_memory_config,
+    )
+    hidden_states = dropout(hidden_states, p=0, training=False)
+    hidden_states = ttnn.add(residual, hidden_states)
+
+    residual = hidden_states
+
+    hidden_states = ttnn.layer_norm(
+        hidden_states,
+        weight=parameters.final_layer_norm.weight,
+        bias=parameters.final_layer_norm.bias,
+        memory_config=whisper_memory_config,
+    )
+
+    hidden_states = ttnn.linear(
+        hidden_states,
+        parameters.fc1.weight,
+        bias=parameters.fc1.bias,
+    )
+
+    hidden_states = ttnn.gelu(hidden_states, memory_config=whisper_memory_config)
+    hidden_states = dropout(hidden_states, p=0, training=False)
+
+    hidden_states = ttnn.linear(
+        hidden_states,
+        parameters.fc2.weight,
+        bias=parameters.fc2.bias,
+        memory_config=whisper_memory_config,
+    )
+
+    hidden_states = dropout(hidden_states, p=0, training=False)
+    hidden_states = ttnn.add(residual, hidden_states)
+
+    # if hidden_states.dtype == torch.float16 and (torch.isinf(hidden_states).any() or torch.isnan(hidden_states).any()):
+    #     clamp_value = torch.finfo(hidden_states.dtype).max - 1000
+    #     hidden_states = torch.clamp(hidden_states, min=-clamp_value, max=clamp_value)
```

Comment:
can this be removed?

Reply:
I've updated the file removing comments.